onnxruntime/contrib_ops/cuda/transformers/beam_search_topk.cu

```diff
@@ -0,0 +1,442 @@
+  BeamSearchOnlineTopKStage1Kernel<T, max_k, kThreadBlockSize>
+      <<<grid, kThreadBlockSize, 0, stream>>>(input, K, vocab_size, (vocab_size + voc_parts - 1) / voc_parts, output_values_tmp, output_indices_tmp);
+
+  LaunchBeamSearchOnlineTopKStage2Kernel<T, max_k>(
+      output_values_tmp,
+      output_indices_tmp,
+      batch_size,
+      num_beams,
+      vocab_size,
+      voc_parts,
+      K,
+      output_values,
+      output_indices,
+      stream);
+}
+
+template <typename T, typename I, int32_t max_k, int32_t thread_block_size>
+__launch_bounds__(thread_block_size) __global__ void BatchTopKKernel(
+    const T* topk_scores,
+    const I* topk_tokens,
+    int32_t* next_indices,
+    int32_t* next_tokens,
+    T* next_scores,
+    int32_t batch_size,
+    int32_t num_beams,
+    int32_t k) {
+  int thread_id = threadIdx.x;
+  int block_id = blockIdx.x;
+  TopK<T, max_k> thread_topk;
+  if (thread_id == 0) {
+    thread_topk.Init();
+
+    int index_block = block_id * num_beams * k;
+    for (int32_t i = 0; i < num_beams * k; i++) {
+      thread_topk.Insert(topk_scores[index_block + i], index_block + i);
+    }
+
+    int index_next = block_id * k;
+    for (int i = 0; i < k; i++) {
+      next_tokens[index_next + i] = topk_tokens[thread_topk.key[i]];
+      next_indices[index_next + i] = (thread_topk.key[i] - index_block) / k;
+      next_scores[index_next + i] = thread_topk.value[i];
+    }
+  }
+}
+
+template <typename T, typename I>
+void LanuchBatchTopKKernel(const T* topk_scores,
+                           const I* topk_tokens,
+                           int32_t* next_indices,
+                           int32_t* next_tokens,
+                           T* next_scores,
+                           int32_t batch_size,
+                           int32_t num_beams,
+                           int32_t k,
+                           cudaStream_t stream) {
+  ORT_ENFORCE(k <= 256, "LanuchBatchTopKKernel doesn't support k >= 256");
+
+#define BatchTopKKernelLauncher(K)                                          \
+  BatchTopKKernel<T, I, K, 32><<<batch_size, 32, 0, stream>>>(topk_scores,  \
+                                                              topk_tokens,  \
+                                                              next_indices, \
+                                                              next_tokens,  \
+                                                              next_scores,  \
+                                                              batch_size,   \
+                                                              num_beams,    \
+                                                              k);
+
+  if (k <= 4) {
+    BatchTopKKernelLauncher(4);
+  } else if (k <= 8) {
+    BatchTopKKernelLauncher(8);
+  } else if (k <= 16) {
+    BatchTopKKernelLauncher(16);
+  } else if (k <= 32) {
+    BatchTopKKernelLauncher(32);
+  } else if (k <= 64) {
+    BatchTopKKernelLauncher(64);
+  } else if (k <= 128) {
```

Comment:
Here we support k=128 and larger. In Launch function,  ORT_ENFORCE(k <= 64, ...), In beam search, I saw k <= 32 limit. It is better to consolidate the limits.

Reply:
They are different: 32 in generation_device_helper.cc is num_beam, which is half of the top K(64) ( k = 2*num_beam). For those case, we call the online topk computation directly. 

If num_beam > 32, we still need to call LaunchBatchTopKKernel. I limit it to 256, which means we only support beam_size <= 128.

Reply:
If we only support beam_size <= 128, we might need add this check somewhere (like in CheckInputs).